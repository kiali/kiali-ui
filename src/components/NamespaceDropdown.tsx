import * as React from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import _ from 'lodash';
import { style } from 'typestyle';
import { Button, Icon, OverlayTrigger, Popover, FormControl, InputGroup } from 'patternfly-react';
import { KialiAppState } from '../store/Store';
import { activeNamespacesSelector, namespaceItemsSelector, namespaceFilterSelector } from '../store/Selectors';
import { KialiAppAction } from '../actions/KialiAppAction';
import { NamespaceActions } from '../actions/NamespaceAction';
import NamespaceThunkActions from '../actions/NamespaceThunkActions';
import Namespace from '../types/Namespace';
import { PfColors } from './Pf/PfColors';
import { HistoryManager, URLParams } from '../app/History';

const namespaceButtonColors = {
  backgroundColor: PfColors.White,
  borderColor: '#18546B',
  color: '#003145'
};

const namespaceButtonStyle = style({
  ...namespaceButtonColors,
  borderRadius: '2px 2px 0 0',
  borderWidth: '1px 1px 3px 1px',
  height: '32px',
  boxSizing: 'border-box',
  padding: '4px 6px 5px 6px',
  // these properties are being overridden by btn:hover/focus and btn-link:hover/focus
  $nest: {
    '&:hover': namespaceButtonColors,
    '&:focus': namespaceButtonColors
  }
});

const namespaceLabelStyle = style({
  fontWeight: 500
});

const namespaceValueStyle = style({
  fontWeight: 600
});

interface NamespaceListType {
  disabled: boolean;
  filter: string;
  activeNamespaces: Namespace[];
  items: Namespace[];
  toggleNamespace: (namespace: Namespace) => void;
  setNamespaces: (namespaces: Namespace[]) => void;
  setFilter: (filter: string) => void;
  refresh: () => void;
  clearAll: () => void;
}

export class NamespaceDropdown extends React.PureComponent<NamespaceListType, {}> {
  constructor(props: NamespaceListType) {
    super(props);
  }

  componentDidMount() {
    this.props.refresh();
    this.syncNamespacesURLParam();
  }

  componentDidUpdate(prevProps: NamespaceListType) {
    if (prevProps.activeNamespaces !== this.props.activeNamespaces) {
      if (this.props.activeNamespaces.length === 0) {
        HistoryManager.deleteParam(URLParams.NAMESPACES);
      } else {
        HistoryManager.setParam(URLParams.NAMESPACES, this.props.activeNamespaces.map(item => item.name).join(','));
      }
    }
  }

  syncNamespacesURLParam = () => {
    const namespaces = (HistoryManager.getParam(URLParams.NAMESPACES) || '').split(',').filter(Boolean);
    if (namespaces.length > 0 && _.difference(namespaces, this.props.activeNamespaces.map(item => item.name))) {
      // We must change the props of namespaces
      const items = namespaces.map(ns => ({ name: ns } as Namespace));
      this.props.setNamespaces(items);
    } else if (namespaces.length === 0 && this.props.activeNamespaces.length !== 0) {
      HistoryManager.setParam(URLParams.NAMESPACES, this.props.activeNamespaces.map(item => item.name).join(','));
    }
  };

  onNamespaceToggled = (a: any) => {
    this.props.toggleNamespace({ name: a.target.value });
  };

  onFilterChange = (event: any) => {
    this.props.setFilter(event.target.value);
  };

  clearFilter = () => {
    this.props.setFilter('');
  };

  namespaceButtonText() {
    if (this.props.activeNamespaces.length === 0) {
      return <span className={namespaceValueStyle}>Select a namespace</span>;
    } else if (this.props.activeNamespaces.length === 1) {
      return (
        <>
          <span className={namespaceLabelStyle}>Namespace:</span>
          <span>&nbsp;</span>
          <span className={namespaceValueStyle}>{this.props.activeNamespaces[0].name}</span>
        </>
      );
    } else {
      return (
        <>
          <span className={namespaceLabelStyle}>Namespaces:</span>
          <span>&nbsp;</span>
          <span className={namespaceValueStyle}>{`${this.props.activeNamespaces.length} namespaces`}</span>
        </>
      );
    }
  }

  getPopoverContent() {
    if (this.props.items.length > 0) {
      const activeMap = this.props.activeNamespaces.reduce((map, namespace) => {
        map[namespace.name] = namespace.name;
        return map;
      }, {});
      const checkboxStyle = style({ marginLeft: 5 });
      const namespaces = this.props.items
        .filter((namespace: Namespace) => namespace.name.includes(this.props.filter))
        .map((namespace: Namespace) => (
          <div id={`namespace-list-item[${namespace.name}]`} key={`namespace-list-item[${namespace.name}]`}>
            <label>
              <input
                type="checkbox"
                value={namespace.name}
                checked={!!activeMap[namespace.name]}
                onChange={this.onNamespaceToggled}
              />
              <span className={checkboxStyle}>{namespace.name}</span>
            </label>
          </div>
        ));

      return (
        <>
          <div>
            <InputGroup>
              <FormControl
                type="text"
                name="namespace-filter"
                placeholder="Filter by keyword..."
                value={this.props.filter}
                onChange={this.onFilterChange}
              />
              {this.props.filter !== '' && (
                <InputGroup.Button>
                  <Button onClick={this.clearFilter}>
                    <Icon name="close" />
                  </Button>
                </InputGroup.Button>
              )}
            </InputGroup>
          </div>
          <div className="text-right">
            <Button disabled={this.props.activeNamespaces.length === 0} bsStyle="link" onClick={this.props.clearAll}>
              Clear all
            </Button>
          </div>
          <div>{namespaces}</div>
        </>
      );
    }
    return <div>No namespaces found or they haven't loaded yet</div>;
  }

  render() {
    const popover = <Popover id="namespace-list-layers-popover">{this.getPopoverContent()}</Popover>;
    return (
      <OverlayTrigger
        onEnter={this.props.refresh}
        overlay={popover}
        placement="bottom"
        trigger={['click']}
        rootClose={true}
      >
        <Button bsClass={`btn btn-link btn-lg  ${namespaceButtonStyle}`} id="namespace-selector">
          {this.namespaceButtonText()} <Icon name="angle-down" />
        </Button>
      </OverlayTrigger>
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    items: namespaceItemsSelector(state)!,
    activeNamespaces: activeNamespacesSelector(state),
    filter: namespaceFilterSelector(state)
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<KialiAppState, void, KialiAppAction>) => {
  return {
    refresh: () => {
      dispatch(NamespaceThunkActions.fetchNamespacesIfNeeded());
    },
    toggleNamespace: (namespace: Namespace) => {
      dispatch(NamespaceActions.toggleActiveNamespace(namespace));
    },
    clearAll: () => {
      dispatch(NamespaceActions.setActiveNamespaces([]));
    },
    setNamespaces: (namespaces: Namespace[]) => {
      dispatch(NamespaceActions.setActiveNamespaces(namespaces));
    },
    setFilter: (filter: string) => {
      dispatch(NamespaceActions.setFilter(filter));
    }
  };
};

const NamespaceDropdownContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(NamespaceDropdown);
export default NamespaceDropdownContainer;
