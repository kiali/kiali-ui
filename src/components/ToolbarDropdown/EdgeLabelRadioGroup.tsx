import * as React from 'react';
import PropTypes from 'prop-types';
import { style } from 'typestyle';
import * as _ from 'lodash';

import { GraphParamsType } from '../../types/Graph';
import { EdgeLabelMode } from '../../types/GraphFilter';

const EDGE_LABEL_MODES = EdgeLabelMode.getValues().reduce((map, edgeLabelMode) => {
  map[edgeLabelMode] = _.capitalize(_.startCase(edgeLabelMode));
  return map;
}, {});

type EdgeLabelRadioGroupProps = {
  graphParams: GraphParamsType;
  onEdgeChanged: PropTypes.func;
};

export class EdgeLabelRadioGroup extends React.Component<EdgeLabelRadioGroupProps> {
  constructor(props: EdgeLabelRadioGroupProps) {
    super(props);
  }

  onEdgeChanged = (event: any) => {
    const selectedEdge = event.target.value;
    this.setState({
      edgeValue: selectedEdge
    });
    this.props.onEdgeChanged(event);
  };

  render() {
    const radioButtonStyle = style({ marginLeft: 5 });

    const edgeItems = Object.keys(EDGE_LABEL_MODES).map((edgeLabelModeKey: any) => (
      <div key={edgeLabelModeKey}>
        <label className={radioButtonStyle}>
          <input
            type="radio"
            name="edge-label-radio-group"
            value={edgeLabelModeKey}
            checked={edgeLabelModeKey === this.props.graphParams.edgeLabelMode}
            onChange={this.onEdgeChanged}
          />
          <span className={radioButtonStyle}>{EDGE_LABEL_MODES[edgeLabelModeKey]}</span>
        </label>
      </div>
    ));

    return (
      <>
        <div>
          <label>Edge Labels:</label>
        </div>
        {edgeItems}
      </>
    );
  }
}

export default EdgeLabelRadioGroup;
