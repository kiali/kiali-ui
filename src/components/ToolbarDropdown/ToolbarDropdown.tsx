import * as React from 'react';
import { DropdownButton, MenuItem } from 'patternfly-react';
import PropTypes from 'prop-types';

type ToolbarDropdownProps = {
  disabled: boolean;
  nameDropdown?: string;
  initialValue: number | string;
  initialLabel: string | undefined;
  onClick: PropTypes.func;
  options: PropTypes.Map<string | number, string>;
};

type ToolbarDropdownState = {
  currentValue: number | string;
  currentName: string | undefined;
};

export class ToolbarDropdown extends React.Component<ToolbarDropdownProps, ToolbarDropdownState> {
  constructor(props: ToolbarDropdownProps) {
    super(props);
    this.state = {
      currentValue: props.initialValue,
      currentName: props.initialLabel
    };
  }

  onKeyChanged = (key: any) => {
    this.setState({ currentValue: key[0], currentName: key[1] });
    this.props.onClick(key[0]);
  };

  render() {
    return (
      <div className="form-group">
        {this.props.nameDropdown && <label style={{ paddingRight: '0.5em' }}>{this.props.nameDropdown}:</label>}
        <DropdownButton title={this.state.currentName} onSelect={this.onKeyChanged}>
          {this.props.options.map(r => (
            <MenuItem key={r[0]} active={r[0] === this.state.currentValue} eventKey={r}>
              {r[1]}
            </MenuItem>
          ))}
        </DropdownButton>
      </div>
    );
  }
}
