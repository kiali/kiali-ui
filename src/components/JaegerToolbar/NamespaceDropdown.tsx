import * as React from 'react';
import { connect } from 'react-redux';
import { KialiAppState } from '../../store/Store';
import Namespace from '../../types/Namespace';
import ToolbarDropdown from '../../components/ToolbarDropdown/ToolbarDropdown';
import { JaegerActions } from '../../actions/JaegerActions';
import { ThunkDispatch } from 'redux-thunk';
import { KialiAppAction } from '../../actions/KialiAppAction';
import NamespaceThunkActions from '../../actions/NamespaceThunkActions';
import { JaegerThunkActions } from '../../actions/JaegerThunkActions';

interface NamespaceDropdownProps {
  disabled: boolean;
  namespace: string;
  items: Namespace[];
  refresh: () => void;
  setNamespace: (service: string) => void;
}

export class NamespaceDropdown extends React.PureComponent<NamespaceDropdownProps, {}> {
  constructor(props: NamespaceDropdownProps) {
    super(props);
  }

  componentDidMount() {
    this.props.refresh();
  }

  render() {
    const { disabled, namespace, setNamespace } = this.props;

    const items: { [key: string]: string } = this.props.items.reduce((list, item) => {
      list[item.name] = item.name;
      return list;
    }, {});

    return (
      <ToolbarDropdown
        id="namespace-selector"
        disabled={disabled || Object.keys(items).length === 0}
        options={items}
        value={namespace}
        label={namespace || 'Select a Namespace'}
        useName={true}
        handleSelect={setNamespace}
      />
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    items: state.namespaces.items,
    disabled: state.namespaces.isFetching,
    namespace: state.jaegerState.search.namespaceSelected
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<KialiAppState, void, KialiAppAction>) => {
  return {
    refresh: () => {
      dispatch(NamespaceThunkActions.fetchNamespacesIfNeeded());
    },
    setNamespace: (namespace: string) => {
      dispatch(JaegerActions.setNamespace(namespace));
      dispatch(JaegerActions.setService(''));
      dispatch(JaegerThunkActions.asyncFetchServices(namespace));
    }
  };
};

const NamespaceDropdownContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(NamespaceDropdown);

export default NamespaceDropdownContainer;
