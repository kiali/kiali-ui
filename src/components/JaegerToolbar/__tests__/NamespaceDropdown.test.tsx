import * as React from 'react';
import { shallow } from 'enzyme';
import { NamespaceDropdown } from '../NamespaceDropdown';
import ToolbarDropdown from '../../../components/ToolbarDropdown/ToolbarDropdown';

describe('NamespaceDropdown', () => {
  let wrapper, refresh, setNamespace;
  beforeEach(() => {
    refresh = jest.fn();
    setNamespace = jest.fn();
    wrapper = shallow(
      <NamespaceDropdown refresh={refresh} setNamespace={setNamespace} namespace={''} disabled={false} items={[]} />
    );
  });

  it('renders NamespaceDropdown correctly without custom', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders NamespaceDropdown correctly with custom', () => {
    wrapper.setProps({ items: [{ name: 'bookinfo' }, { name: 'istio-system' }] });
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('disable ToolbarDropwdown if no namespaces', () => {
    expect(wrapper.find(ToolbarDropdown).length).toEqual(1);
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeTruthy();
    wrapper.setProps({ items: [{ name: 'bookinfo' }] });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeFalsy();
    wrapper.setProps({ disabled: true });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeTruthy();
  });

  it('set value if namespace selected', () => {
    const ns = 'bookinfo';
    wrapper.setProps({ namespace: ns });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['value']
    ).toBe(ns);
  });

  it('set label if namespace selected', () => {
    const ns = 'bookinfo';
    const label = 'Select a Namespace';
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['label']
    ).toBe(label);
    wrapper.setProps({ namespace: ns });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['label']
    ).toBe(ns);
  });

  it('set items in options in dropdown', () => {
    const namespaces = [{ name: 'bookinfo' }, { name: 'istio-system' }];
    const items: { [key: string]: string } = namespaces.reduce((list, item) => {
      list[item.name] = item.name;
      return list;
    }, {});
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['options']
    ).toEqual({});
    wrapper.setProps({ items: namespaces });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['options']
    ).toEqual(items);
  });

  it('set handleSelect in dropdown', () => {
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['handleSelect']
    ).toBe(setNamespace);
  });
});
