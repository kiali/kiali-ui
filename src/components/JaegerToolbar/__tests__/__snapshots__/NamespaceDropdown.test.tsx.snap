// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceDropdown renders NamespaceDropdown correctly with custom 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NamespaceDropdown
    disabled={false}
    items={
      Array [
        Object {
          "name": "bookinfo",
        },
        Object {
          "name": "istio-system",
        },
      ]
    }
    namespace=""
    refresh={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    setNamespace={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "disabled": false,
      "handleSelect": [MockFunction],
      "id": "namespace-selector",
      "label": "Select a Namespace",
      "options": Object {
        "bookinfo": "bookinfo",
        "istio-system": "istio-system",
      },
      "useName": true,
      "value": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "disabled": false,
        "handleSelect": [MockFunction],
        "id": "namespace-selector",
        "label": "Select a Namespace",
        "options": Object {
          "bookinfo": "bookinfo",
          "istio-system": "istio-system",
        },
        "useName": true,
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`NamespaceDropdown renders NamespaceDropdown correctly without custom 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NamespaceDropdown
    disabled={false}
    items={Array []}
    namespace=""
    refresh={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    setNamespace={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "disabled": true,
      "handleSelect": [MockFunction],
      "id": "namespace-selector",
      "label": "Select a Namespace",
      "options": Object {},
      "useName": true,
      "value": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "disabled": true,
        "handleSelect": [MockFunction],
        "id": "namespace-selector",
        "label": "Select a Namespace",
        "options": Object {},
        "useName": true,
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
